--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/dbus_systemcalc.py-v3.55.orig	2025-04-03 07:27:50
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/dbus_systemcalc.py-v3.55	2025-04-07 16:11:03
@@ -1,6 +1,8 @@
 #!/usr/bin/python3 -u
 # -*- coding: utf-8 -*-
 
+#### modified for GuiMods
+
 from dbus.mainloop.glib import DBusGMainLoop
 import dbus
 import argparse
@@ -72,6 +74,20 @@
 				'/Ac/Out/L1/I': dummy,
 				'/Ac/Out/L2/I': dummy,
 				'/Ac/Out/L3/I': dummy,
+#### add for GuiMods
+				'/Ac/Out/L1/V': dummy,
+				'/Ac/Out/L2/V': dummy,
+				'/Ac/Out/L3/V': dummy,
+				'/Ac/Out/L1/F': dummy,
+				'/Ac/Out/L2/F': dummy,
+				'/Ac/Out/L3/F': dummy,
+				'/Ac/ActiveIn/L1/V': dummy,
+				'/Ac/ActiveIn/L2/V': dummy,
+				'/Ac/ActiveIn/L3/V': dummy,
+				'/Ac/ActiveIn/L1/F': dummy,
+				'/Ac/ActiveIn/L2/F': dummy,
+				'/Ac/ActiveIn/L3/F': dummy,
+
 				'/Connected': dummy,
 				'/ProductId': dummy,
 				'/ProductName': dummy,
@@ -109,7 +125,14 @@
 				'/Ac/L3/Power': dummy,
 				'/Ac/L1/Current': dummy,
 				'/Ac/L2/Current': dummy,
-				'/Ac/L3/Current': dummy},
+				'/Ac/L3/Current': dummy,
+#### add for GuiMods
+				'/Ac/L1/Voltage': dummy,
+				'/Ac/L2/Voltage': dummy,
+				'/Ac/L3/Voltage': dummy,
+				'/Ac/L1/Frequency': dummy,
+				'/Ac/L2/Frequency': dummy,
+				'/Ac/L3/Frequency': dummy},
 			'com.victronenergy.genset' : {
 				'/Connected': dummy,
 				'/ProductName': dummy,
@@ -122,6 +145,14 @@
 				'/Ac/L1/Current': dummy,
 				'/Ac/L2/Current': dummy,
 				'/Ac/L3/Current': dummy,
+#### add for GuiMods
+				'/Ac/L1/Voltage': dummy,
+				'/Ac/L2/Voltage': dummy,
+				'/Ac/L3/Voltage': dummy,
+				'/Ac/L1/Frequency': dummy,
+				'/Ac/L2/Frequency': dummy,
+				'/Ac/L3/Frequency': dummy,
+
 				'/StarterVoltage': dummy},
 			'com.victronenergy.settings' : {
 				'/Settings/SystemSetup/AcInput1' : dummy,
@@ -143,6 +174,9 @@
 				'/Ac/Out/L1/S': dummy,
 				'/Ac/Out/L1/V': dummy,
 				'/Ac/Out/L1/I': dummy,
+#### add for GuiMods
+				'/Ac/Out/L1/F': dummy,
+
 				'/Ac/Out/L2/P': dummy,
 				'/Ac/Out/L2/S': dummy,
 				'/Ac/Out/L2/V': dummy,
@@ -170,6 +204,9 @@
 				'/Ac/Out/L1/P': dummy,
 				'/Ac/Out/L1/V': dummy,
 				'/Ac/Out/L1/I': dummy,
+#### add for GuiMods
+				'/Ac/Out/L1/F': dummy,
+
 				'/Ac/In/1/L2/P': dummy,
 				'/Ac/In/1/L2/I': dummy,
 				'/Ac/In/2/L2/P': dummy,
@@ -188,11 +225,19 @@
 				'/Soc': dummy},
 			'com.victronenergy.dcsystem': {
 				'/Dc/0/Voltage': dummy,
-				'/Dc/0/Power': dummy,
-				'/Dc/0/Current': dummy,
+				'/Dc/0/Power': dummy
 			},
 			'com.victronenergy.alternator': {
 				'/Dc/0/Power': dummy
+			},
+#### added for GuiMods
+			'com.victronenergy.dcsource': {
+				'/Dc/0/Power': dummy,
+				'/Settings/MonitorMode': dummy
+			},
+			'com.victronenergy.motordrive':
+			{
+				'/Dc/0/Power': dummy
 			}
 		}
 
@@ -346,6 +391,33 @@
 			'/Ac/ActiveIn/L2/Current': {'gettext': '%.1F A'},
 			'/Ac/ActiveIn/L3/Current': {'gettext': '%.1F A'},
 			'/Ac/ActiveIn/NumberOfPhases': {'gettext': '%d'},
+#### added for GuiMods
+			'/Dc/WindGenerator/Power': {'gettext': '%.0F W'},
+			'/Dc/MotorDrive/Power': {'gettext': '%.0F W'},
+			'/Ac/Grid/L1/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Grid/L2/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Grid/L3/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Grid/Frequency': {'gettext': '%.1F Hz'},
+			'/Ac/Genset/L1/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Genset/L2/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Genset/L3/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Genset/Frequency': {'gettext': '%.1F Hz'},
+			'/Ac/ConsumptionOnOutput/L1/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ConsumptionOnOutput/L2/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ConsumptionOnOutput/L3/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ConsumptionOnOutput/Frequency': {'gettext': '%.1F Hz'},
+			'/Ac/ConsumptionOnInput/L1/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ConsumptionOnInput/L2/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ConsumptionOnInput/L3/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ConsumptionOnInput/Frequency': {'gettext': '%.1F Hz'},
+			'/Ac/Consumption/L1/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Consumption/L2/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Consumption/L3/Voltage': {'gettext': '%.1F V'},
+			'/Ac/Consumption/Frequency': {'gettext': '%.1F Hz'},
+			'/Ac/ActiveIn/L1/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ActiveIn/L2/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ActiveIn/L3/Voltage': {'gettext': '%.1F V'},
+			'/Ac/ActiveIn/Frequency': {'gettext': '%.1F Hz'},
 		}
 
 		for m in self._modules:
@@ -364,6 +436,9 @@
 		self._changed = True
 		for service, instance in self._dbusmonitor.get_service_list().items():
 			self._device_added(service, instance, do_service_change=False)
+
+#### added for GuiMods
+		self.dcSystemPower = [0, 0, 0]
 
 		self._handleservicechange()
 		self._updatevalues()
@@ -624,6 +699,35 @@
 				newvalues['/Dc/Alternator/Power'] = p
 			else:
 				newvalues['/Dc/Alternator/Power'] += p
+
+
+#### added for GuiMods
+		# ==== MOTOR DRIVE ====
+		motordrives = self._dbusmonitor.get_service_list('com.victronenergy.motordrive')
+		for motordrive in motordrives:
+			p = self._dbusmonitor.get_value(motordrive, '/Dc/0/Power')
+			if p is None:
+				p = 0
+
+			if '/Dc/MotorDrive/Power' not in newvalues:
+				newvalues['/Dc/MotorDrive/Power'] = p
+			else:
+				newvalues['/Dc/MotorDrive/Power'] += p
+
+#### added for GuiMods
+		# ==== DC SOURCES ====
+		dcSources = self._dbusmonitor.get_service_list('com.victronenergy.dcsource')
+		for dcSource in dcSources:
+			monitorMode = self._dbusmonitor.get_value(dcSource,'/Settings/MonitorMode')
+			# ==== WIND GENERATOR ====
+			if monitorMode == -8:
+				p = self._dbusmonitor.get_value(dcSource, '/Dc/0/Power')
+				if p is None:
+					continue
+				if '/Dc/WindGenerator/Power' not in newvalues:
+					newvalues['/Dc/WindGenerator/Power'] = p
+				else:
+					newvalues['/Dc/WindGenerator/Power'] += p
 
 		# ==== CHARGERS ====
 		chargers = self._dbusmonitor.get_service_list('com.victronenergy.charger')
@@ -789,6 +893,9 @@
 				charger_power = newvalues.get('/Dc/Charger/Power', 0)
 				fuelcell_power = newvalues.get('/Dc/FuelCell/Power', 0)
 				alternator_power = newvalues.get('/Dc/Alternator/Power', 0)
+#### added for GuiMods
+				windgen_power = newvalues.get('/Dc/WindGenerator/Power', 0)
+				motordrive_power = newvalues.get('/Dc/MotorDrive/Power', 0)
 
 				# If there are VE.Direct inverters, remove their power from the
 				# DC estimate. This is done using the AC value when the DC
@@ -804,7 +911,14 @@
 							i, '/Ac/Out/L1/V', 0) * self._dbusmonitor.get_value(
 							i, '/Ac/Out/L1/I', 0)
 				newvalues['/Dc/System/MeasurementType'] = 0 # estimated
-				newvalues['/Dc/System/Power'] = dc_pv_power + charger_power + fuelcell_power + alternator_power + vebuspower + inverter_power - battery_power
+#### changed for GuiMods
+				# average DC system power over 3 passes (seconds) to minimize wild swings in displayed value
+				self.dcSystemPower[2] = self.dcSystemPower[1]
+				self.dcSystemPower[1] = self.dcSystemPower[0]
+#### changed for GuiMods - include wind and motor drive
+				#### was newvalues['/Dc/System/Power'] = dc_pv_power + charger_power + fuelcell_power + alternator_power + vebuspower + inverter_power - battery_power
+				self.dcSystemPower[0] = dc_pv_power + charger_power + fuelcell_power + alternator_power + vebuspower + inverter_power - battery_power + windgen_power - motordrive_power
+				newvalues['/Dc/System/Power'] = (self.dcSystemPower[0] + self.dcSystemPower[1] + self.dcSystemPower[2]) / 3
 				try:
 					newvalues['/Dc/System/Current'] = \
 						newvalues['/Dc/System/Power'] / newvalues['/Dc/Battery/Voltage']
@@ -861,6 +975,13 @@
 
 		consumption = { "L1" : None, "L2" : None, "L3" : None }
 		currentconsumption = { "L1" : None, "L2" : None, "L3" : None }
+
+#### added for GuiMods
+		voltageIn = { "L1" : None, "L2" : None, "L3" : None }
+		voltageOut = { "L1" : None, "L2" : None, "L3" : None }
+		frequencyIn = None
+		frequencyOut = None
+
 		for device_type, em, _types in (('Grid', grid_meter, (1, 3)), ('Genset', genset_meter, (2,))):
 			# If a grid meter is present we use values from it. If not, we look at the multi. If it has
 			# AcIn1 or AcIn2 connected to the grid, we use those values.
@@ -876,6 +997,12 @@
 				if em is not None:
 					p = self._dbusmonitor.get_value(em.service, '/Ac/%s/Power' % phase)
 					mc = self._dbusmonitor.get_value(em.service, '/Ac/%s/Current' % phase)
+#### added for GuiMods
+					if voltageIn[phase] == None:
+						voltageIn[phase] = self._dbusmonitor.get_value(em.service, '/Ac/%s/Voltage' % phase)
+					if frequencyIn == None:
+						frequencyIn = self._dbusmonitor.get_value(em.service, '/Ac/%s/Frequency' % phase)
+
 					# Compute consumption between energy meter and multi (meter power - multi AC in) and
 					# add an optional PV inverter on input to the mix.
 					c = None
@@ -885,6 +1012,12 @@
 							try:
 								c = _safeadd(c, -self._dbusmonitor.get_value(multi_path, '/Ac/ActiveIn/%s/P' % phase))
 								cc = _safeadd(cc, -self._dbusmonitor.get_value(multi_path, '/Ac/ActiveIn/%s/I' % phase))
+#### added for GuiMods
+								if voltageIn[phase] == None:
+									voltageIn[phase] = self._dbusmonitor.get_value(em.service, '/Ac/ActiveIn/%s/V' % phase)
+								if frequencyIn == None:
+									frequencyIn = self._dbusmonitor.get_value(em.service, '/Ac/ActiveIn/%s/F' % phase)
+
 							except TypeError:
 								pass
 						elif non_vebus_inverter is not None and active_input in (0, 1):
@@ -892,6 +1025,12 @@
 								try:
 									c = _safeadd(c, -self._dbusmonitor.get_value(i, '/Ac/In/%d/%s/P' % (active_input+1, phase)))
 									cc = _safeadd(cc, -self._dbusmonitor.get_value(i, '/Ac/In/%d/%s/I' % (active_input+1, phase)))
+#### added for GuiMods
+									if voltageIn[phase] == None:
+										voltageIn[phase] = self._dbusmonitor.get_value(i, '/Ac/In/%d/%s/V' % (active_input+1, phase))
+									if frequencyIn == None:
+										frequencyIn = self._dbusmonitor.get_value(i, '/Ac/In/%d/%s/F' % (active_input+1, phase))
+
 								except TypeError:
 									pass
 
@@ -909,12 +1048,26 @@
 							consumption[phase] = _safeadd(0, consumption[phase])
 							currentconsumption[phase] = _safeadd(0, currentconsumption[phase])
 							mc = self._dbusmonitor.get_value(multi_path, '/Ac/ActiveIn/%s/I' % phase)
+#### added for GuiMods
+							if voltageIn[phase] == None:
+								voltageIn[phase] = self._dbusmonitor.get_value(multi_path, '/Ac/ActiveIn/%s/V' % phase)
+							if frequencyIn == None:
+								freq = self._dbusmonitor.get_value(multi_path, '/Ac/ActiveIn/%s/F' % phase)
+								if freq != None:
+									frequencyIn = freq
+
 						elif non_vebus_inverter is not None and active_input in (0, 1):
 							for i in non_vebus_inverters:
 								p = _safeadd(p,
 									self._dbusmonitor.get_value(i, '/Ac/In/%d/%s/P' % (active_input + 1, phase)))
 								mc = _safeadd(mc,
 									self._dbusmonitor.get_value(i, '/Ac/In/%d/%s/I' % (active_input + 1, phase)))
+#### added for GuiMods
+								if voltageIn[phase] == None:
+									voltageIn[phase] = self._dbusmonitor.get_value(i, '/Ac/In/%d/%s/V' % (active_input + 1, phase))
+								if frequencyIn == None:
+									frequencyIn = self._dbusmonitor.get_value(i, '/Ac/In/%d/%s/F' % (active_input + 1, phase))
+
 							if p is not None:
 								consumption[phase] = _safeadd(0, consumption[phase])
 								currentconsumption[phase] = _safeadd(0, currentconsumption[phase])
@@ -929,9 +1082,18 @@
 
 				newvalues['/Ac/%s/%s/Power' % (device_type, phase)] = p
 				newvalues['/Ac/%s/%s/Current' % (device_type, phase)] = mc
+#### added for GuiMods
+				if p != None:
+					newvalues['/Ac/%s/%s/Voltage' % (device_type, phase)] = voltageIn[phase]
+					newvalues['/Ac/%s/Frequency' % (device_type)] = frequencyIn
+
 				if ac_in_guess in _types:
 					newvalues['/Ac/ActiveIn/%s/Power' % (phase,)] = p
 					newvalues['/Ac/ActiveIn/%s/Current' % (phase,)] = mc
+#### added for GuiMods
+					if p != None:
+						newvalues['/Ac/ActiveIn/%s/Voltage' % (phase,)] = voltageIn[phase]
+						newvalues['/Ac/ActiveIn/Frequency'] = frequencyIn
 
 			self._compute_number_of_phases('/Ac/%s' % device_type, newvalues)
 			self._compute_number_of_phases('/Ac/ActiveIn', newvalues)
@@ -968,16 +1130,28 @@
 			if use_ac_out:
 				c = newvalues.get('/Ac/PvOnOutput/%s/Power' % phase)
 				a = newvalues.get('/Ac/PvOnOutput/%s/Current' % phase)
+#### added for GuiMods
+				if voltageOut[phase] == None:
+					voltageOut[phase] = newvalues.get('/Ac/PvOnOutput/%s/Voltage' % phase)
+				if frequencyOut == None:
+					frequencyOut = newvalues.get('/Ac/PvOnOutput/%s/Frequency' % phase)
+
 				if multi_path is None:
 					for inv in non_vebus_inverters:
 						ac_out = self._dbusmonitor.get_value(inv, '/Ac/Out/%s/P' % phase)
 						i = self._dbusmonitor.get_value(inv, '/Ac/Out/%s/I' % phase)
+#### added for GuiMods
+						if voltageOut[phase] == None:
+							voltageOut[phase] = self._dbusmonitor.get_value(inv, '/Ac/Out/%s/V' % phase)
+						if frequencyOut == None:
+							frequencyOut = self._dbusmonitor.get_value(inv, '/Ac/Out/%s/F' % phase)
 
 						# Some models don't show power, try apparent power,
 						# else calculate it
 						if ac_out is None:
 							ac_out = self._dbusmonitor.get_value(inv, '/Ac/Out/%s/S' % phase)
 							if ac_out is None:
+#### modified for GuiMods
 								u = self._dbusmonitor.get_value(inv, '/Ac/Out/%s/V' % phase)
 								if None not in (i, u):
 									ac_out = i * u
@@ -988,6 +1162,11 @@
 					c = _safeadd(c, ac_out)
 					i_out = self._dbusmonitor.get_value(multi_path, '/Ac/Out/%s/I' % phase)
 					a = _safeadd(a, i_out)
+#### added for GuiMods
+					if voltageOut[phase] == None:
+						voltageOut[phase] = self._dbusmonitor.get_value(multi_path, '/Ac/Out/%s/V' % phase)
+					if frequencyOut == None:
+						frequencyOut = self._dbusmonitor.get_value(multi_path, '/Ac/Out/%s/F' % phase)
 				c = _safemax(0, c)
 				a = _safemax(0, a)
 			newvalues['/Ac/ConsumptionOnOutput/%s/Power' % phase] = c
@@ -997,6 +1176,21 @@
 			if has_ac_in_system:
 				newvalues['/Ac/ConsumptionOnInput/%s/Power' % phase] = consumption[phase]
 				newvalues['/Ac/ConsumptionOnInput/%s/Current' % phase] = currentconsumption[phase]
+#### added for GuiMods
+			newvalues['/Ac/ConsumptionOnOutput/%s/Voltage' % phase] = voltageOut[phase]
+			newvalues['/Ac/ConsumptionOnInput/%s/Voltage' % phase] = voltageIn[phase]
+			if voltageOut[phase] != None:
+				newvalues['/Ac/Consumption/%s/Voltage' % phase] = voltageOut[phase]
+			elif voltageIn[phase] != None:
+				newvalues['/Ac/Consumption/%s/Voltage' % phase] = voltageIn[phase]
+			if frequencyIn != None:
+				newvalues['/Ac/ConsumptionOnInput/Frequency'] = frequencyIn
+			if frequencyOut != None:
+				newvalues['/Ac/ConsumptionOnOutput/Frequency'] = frequencyOut
+			if frequencyOut != None:
+				newvalues['/Ac/Consumption/Frequency'] = frequencyOut
+			elif frequencyIn != None:
+				newvalues['/Ac/Consumption/Frequency'] = frequencyIn
 
 		self._compute_number_of_phases('/Ac/Consumption', newvalues)
 		self._compute_number_of_phases('/Ac/ConsumptionOnOutput', newvalues)
