--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/dbus_generator.py.orig	2024-07-18 07:51:43
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/dbus_generator.py	2024-05-14 07:00:12
@@ -1,5 +1,15 @@
 #!/usr/bin/python3 -u
 # -*- coding: utf-8 -*-
+
+#### GuiMods
+#### This file has been modified to allow the generator running state derived from the generator digital input
+####	or the genset AC input
+#### If the incoming generator state changes, the manual start state is updated
+#### time accumulation is suspended when the generator is not running
+#### A switch in the generator settings menucontrols whethter the incoming state affects manual start or time accumulaiton
+#### It is now possible to start the generator manually and have it stop automatically based on the preset conditions
+####	for automaitc start / stop
+#### Search for #### GuiMods to find changes
 
 from dbus.mainloop.glib import DBusGMainLoop
 import dbus
@@ -81,7 +91,10 @@
 				'/VebusService': dummy,
 				'/Dc/Battery/Voltage': dummy,
 				'/Dc/Battery/Current': dummy,
-				'/Dc/Battery/Soc': dummy
+				'/Dc/Battery/Soc': dummy,
+#### GuiMods
+				'/Ac/Genset/Frequency': dummy,
+				'/Ac/In/NumberOfAcInputs': dummy
 				}
 			}
 
@@ -93,6 +106,9 @@
 			'accumulateddaily': ['/Settings/{0}/AccumulatedDaily', '', 0, 0, True],
 			'accumulatedtotal': ['/Settings/{0}/AccumulatedTotal', 0, 0, 0, True], # Internal, can't be reset by the user
 			'accumulatedtotalOffset': ['/Settings/{0}/AccumulatedTotalOffset', 0, 0, 0], # For calculating user runtime
+#### GuiMods
+			'linkManualStartToExternal': ['/Settings/{0}/LinkToExternalStatus', 0, 0, 0, True],
+
 			'batterymeasurement': ['/Settings/{0}/BatteryService', 'default', 0, 0],
 			'minimumruntime': ['/Settings/{0}/MinimumRuntime', 0, 0, 86400],  # minutes
 			'stoponac1enabled': ['/Settings/{0}/StopWhenAc1Available', 0, 0, 1],
