--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/MbEditBoxDateTime.qml.orig	2024-07-09 10:19:45
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/MbEditBoxDateTime.qml	2024-05-14 07:00:12
@@ -3,6 +3,10 @@
 MbEditBox {
 	id: root

+////// GuiMods — DarkMode
+	property VBusItem darkModeItem: VBusItem { bind: "com.victronenergy.settings/Settings/Gui/ColorScheme" }
+	property bool darkMode: darkModeItem.valid && darkModeItem.value == 0
+
 	property string format: "yyyy-MM-dd hh:mm"
 	property bool utc: false

@@ -12,7 +16,8 @@
 	overwriteMode: true
 	numericOnlyLayout: true
 	upDownText: qsTr("Select number")
-	textInput.color: editMode ? (vePlatform.secondsFromString(_editText, format) !== -1 ? "black" : "red") : "black"
+////// GuiMods — DarkMode
+	textInput.color: editMode ? (vePlatform.secondsFromString(_editText, format) !== -1 ? (!darkMode ? "#000000" : "#fdfdfd") : "red") : (!darkMode ? "#000000" : "#fdfdfd")

 	// note: overwritten by MbEditBoxTime!
 	function getTimeSeconds(str) {
