--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/MbItemDigitalInput.qml-v3.40.orig	2024-07-09 10:19:45
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/MbItemDigitalInput.qml-v3.40	2025-02-02 17:42:41
@@ -14,8 +14,11 @@
 		MbOption { description: qsTr("Smoke alarm"); value: 6 },
 		MbOption { description: qsTr("Fire alarm"); value: 7 },
 		MbOption { description: qsTr("CO2 alarm"); value: 8 },
+		MbOption { description: qsTr("Generic I/O"); value: 10 },
 		MbOption { description: qsTr("Generator"); value: 9 },
-		MbOption { description: qsTr("Touch input control"); value: 11 }
+		MbOption { description: qsTr("Touch input control"); value: 11 },
+//// added for ExtTransferSwitch package
+		MbOption { description: qsTr("External transfer switch"); value: 12 }
 	]
 	onValueChanged: {
 		if (valid) {
