--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/MbSpinBox.qml-v2.89.orig	2024-07-09 10:19:46
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/MbSpinBox.qml-v2.89	2023-05-19 11:20:34
@@ -1,3 +1,5 @@
+////// GuiMods — added DarkMode
+
 import QtQuick 1.1
 import Qt.labs.components.native 1.0
 import com.victron.velib 1.0
@@ -6,6 +8,10 @@
 	id: root
 	cornerMark: !readOnly && !spinbox.enabled
 
+////// GuiMods — DarkMode
+	property VBusItem darkModeItem: VBusItem { bind: "com.victronenergy.settings/Settings/GuiMods/DarkMode" }
+	property bool darkMode: darkModeItem.valid && darkModeItem.value == 1
+
 	property bool valid: vItem.value !== undefined
 	property alias bind: vItem.bind
 	property string description
@@ -82,10 +88,12 @@
 
 		MbBackgroundRect {
 			id: graytag
-			color: !spinbox.enabled? "#ddd": "#fff"
+////// GuiMods — DarkMode
+			color: !darkMode ? (!spinbox.enabled ? "#ddd": "#fff") : (!spinbox.enabled ? "#4b4b4b": "#747474")
 			height: spinbox.height + 6
 			width:  spinbox.width  + unit.width + 10
-			border.color: "#ddd"
+////// GuiMods — DarkMode
+			border.color: !darkMode ? "#ddd" : "#4b4b4b"
 			border.width: spinbox.enabled ? 1 : 0
 			anchors {
 				right: parent.right; rightMargin: style.marginDefault
