--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/MbSubMenu.qml.orig	2024-07-09 10:19:47
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/MbSubMenu.qml	2024-05-14 07:00:12
@@ -6,6 +6,10 @@
 	id: root
 	width: pageStack ? pageStack.currentPage.width : 0
 
+////// GuiMods — DarkMode
+	property VBusItem darkModeItem: VBusItem { bind: "com.victronenergy.settings/Settings/GuiMods/DarkMode" }
+	property bool darkMode: darkModeItem.valid && darkModeItem.value == 1
+
 	property string description
 	property VBusItem item: VBusItem {}
 	property string iconId: "icon-toolbar-enter"
@@ -59,6 +63,7 @@
 			right: root.right; rightMargin: style.marginDefault
 			verticalCenter: parent.verticalCenter
 		}
-		iconId: root.iconId ? root.iconId + (root.ListView.isCurrentItem ? "-active" : "") : ""
+////// GuiMods — DarkMode
+		iconId: root.iconId ? root.iconId + (root.ListView.isCurrentItem || darkMode ? "-active" : "") : ""
 	}
 }
