--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/OverviewConnection.qml.orig	2024-07-09 10:19:49
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/OverviewConnection.qml	2024-05-14 07:00:12
@@ -22,6 +22,10 @@
 Item {
 	id: root

+////// GuiMods — DarkMode
+	property VBusItem darkModeItem: VBusItem { bind: "com.victronenergy.settings/Settings/Gui/ColorScheme" }
+	property bool darkMode: darkModeItem.valid && darkModeItem.value == 0
+
 	property Path path: emptyPath

 	property Path straight: Path {
@@ -37,9 +41,11 @@

 	property bool active
 	property int ballCount: 4
-	property color ballColor: "#4789d0"
+////// GuiMods — DarkMode
+	property color ballColor: !darkMode ? "#4789d0" : "#386ca5"
 	property real ballDiameter: lineWidth * 2 + 1
-	property color lineColor: "#4789d0"
+////// GuiMods — DarkMode
+	property color lineColor: !darkMode ? "#4789d0" : "#386ca5"
 	property int lineWidth: 3
 	property int value
 	property bool startPointVisible: true
