--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/PageSettingsDisplay.qml-v3.34.orig	2024-07-09 10:19:58
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/PageSettingsDisplay.qml-v3.34	2025-02-02 17:44:10
@@ -65,6 +65,15 @@
 			name: qsTr("Show tanks overview")
 		}
 
+//////// add Gui Mods menu
+		MbSubMenu {
+			id: guiModsMenu
+			description: qsTr("Gui Mods")
+			subpage: Component {
+				PageSettingsGuiMods { }
+			}
+		}
+
 		MbItemOptions {
 			id: languageSelect
 			description: qsTr("Language")
