--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/PageMain.qml-v3.60.orig	2018-03-09 04:34:56
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/PageMain.qml-v3.60	2025-06-25 11:28:41
@@ -1,11 +1,46 @@
 import QtQuick 2
+import "utils.js" as Utils
 import com.victron.velib 1.0
 
 MbPage {
 	id: root
 	title: qsTr("Device List")
 
+//////// GuiMods put Settings, Notifications, Remove disconnected... at top of list
+	property VBusItem moveSettings: VBusItem { id: moveSettings; bind: Utils.path("com.victronenergy.settings", "/Settings/GuiMods/MoveSettings")}
+	property bool settingsAtTop: moveSettings.valid && moveSettings.value === 1
+
 	model: VisualModels {
+//////// GuiMods put Settings, Notifications, Remove disconnected... at top of list
+		VisibleItemModel { //////// use VisualItemModel below for v2.93 and earlier
+			MbSubMenu {
+				description: qsTr("Settings")
+				subpage: Component { PageSettings {} }
+				show: settingsAtTop
+			}
+
+			MbSubMenu {
+				id: menuNotificationsTop
+				description: qsTr("Notifications")
+				item: VBusItem { value: menuNotifications.subpage.summary }
+				subpage: PageNotifications { }
+				show: settingsAtTop
+			}
+
+			MbOK {
+				description: qsTr("Remove disconnected devices")
+				value: qsTr("Press to remove")
+				show: settingsAtTop && deviceList.disconnectedDevices != 0
+				editable: true
+
+				onClicked: {
+					listview.decrementCurrentIndex()
+					deviceList.removeDisconnected()
+				}
+			}
+		}
+//////// end GuiMods put Settings, Notifications, Remove disconnected... at top of list
+
 		DelegateModel {
 			model: VeSortFilterProxyModel {
 				model: DeviceList {
@@ -33,17 +68,22 @@
 				description: qsTr("Notifications")
 				item: VBusItem { value: menuNotifications.subpage.summary }
 				subpage: PageNotifications { }
+//////// GuiMods hide this if added at top
+				show: !settingsAtTop
 			}
 
 			MbSubMenu {
 				description: qsTr("Settings")
 				subpage: Component { PageSettings {} }
+//////// GuiMods hide this if added at top
+				show: !settingsAtTop
 			}
 
 			MbOK {
 				description: qsTr("Remove disconnected devices")
 				value: qsTr("Press to remove")
-				show: deviceList.disconnectedDevices != 0
+//////// GuiMods hide this if added at top
+				show: !settingsAtTop && deviceList.disconnectedDevices != 0
 				editable: true
 
 				onClicked: {
