--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/PageSettingsGenerator.qml.orig	2025-01-12 08:33:22
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/PageSettingsGenerator.qml	2025-02-02 17:44:35
@@ -66,7 +66,18 @@
 			}
 		}
 
+//// GuiMods
 		MbSwitch {
+			name: qsTr("Link to external running state")
+			bind: Utils.path(settingsBindPrefix, "/LinkToExternalStatus")
+			onClicked:
+			{
+				if (!checked)
+					toast.createToast(qsTr("Manual run will be synchronized with the generaror 'is running digital input' or AC input"), 10000, "icon-info-active")
+			}
+		}
+
+		MbSwitch {
 			name: qsTr("Alarm when generator is not in auto start mode")
 			bind: Utils.path(settingsBindPrefix, "/Alarms/AutoStartDisabled")
 			onClicked: {
