--- /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/Tile.qml.orig	2024-07-09 10:20:00
+++ /Users/Kevin/GitHub/GuiMods.copy/FileSets/PatchSource/Tile.qml	2024-05-14 07:00:12
@@ -3,9 +3,14 @@
 Rectangle {
 	id: root
 
+////// GuiMods — DarkMode
+	property VBusItem darkModeItem: VBusItem { bind: "com.victronenergy.settings/Settings/GuiMods/DarkMode" }
+	property bool darkMode: darkModeItem.valid && darkModeItem.value == 1
+
 	color: "#009ec6"
 	border.width: 2
-	border.color: "#fff"
+////// GuiMods — DarkMode
+	border.color: !darkMode ? "#fff" : "#202020"
 	clip: true
 
 	property string title
@@ -21,7 +26,8 @@
 		id: titleField
 		font.pixelSize: 13
 		text: title
-		color: "white"
+////// GuiMods — DarkMode
+		color: !darkMode ? "white" : "#ddd"
 		height: text === "" ? 0 : paintedHeight
 		anchors {
 			top: parent.top; topMargin: 5
@@ -34,7 +40,8 @@
 		width: parent.width - 10
 		height: 1
 		visible: title !== ""
-		color: "white"
+////// GuiMods — DarkMode
+		color: !darkMode ? "white" : "#ddd"
 		anchors {
 			top: titleField.bottom
 			left: titleField.left
